Step1 리뷰 요청드립니다.

ljhokgo opened the pull request:
안녕하세요, 이번에 리뷰를 받게된 이재훈이라고 합니다.
잘 부탁드립니다.

문자열 덧셈 계산기 구현해 보았습니다.
리뷰 부탁드립니다.

구현하다 고민되는 부분이 있어 질문 드립니다.

1. StringAddCalculator 클래스 내 splitWithCustomDelimiter() 에서 Custom Delimiter를 가져오는 부분과 split 하는 부분을 분리하고 싶은데요.. 
찾아낸 custom delimiter와 m.group(2)를 리턴해야 할 것 같은데.. 분리하는 것이 오히려 비효율적일까요?

2. Numbers 클래스 내 createNumbers() 에서 numbers 리스트 초기화 시 지금 코드처럼 하는 것이 좋을지 주석 처리한 람다로 처리하는 것이 좋을지 궁금합니다.

sah3122 commented on src/main/java/study/calculator/StringAddCalculator.java:
```diff
@@ -0,0 +1,39 @@
+            return m.group(2).split(customDelimiter);
+        }
+
+        return null;
```
null을 반환하는것은 좋지 않은 패턴입니다. 
Custom구분자가 아닐경우 Default구분자로 분리하는 로직을 null대신 사용해보세요!

sah3122 commented on src/test/java/study/calculator/StringAddCalculatorTest.java:
```diff
@@ -0,0 +1,47 @@
+
+public class StringAddCalculatorTest {
+    @Test
+    public void splitAndSum_null_또는_빈문자() {
```
@ParameterizedTest
@NullAndEmptySource
의 용도를 찾아보고 사용해보는건 어떨까요 ? 
https://junit.org/junit5/docs/current/user-guide/

sah3122 commented on src/test/java/study/calculator/StringAddCalculatorTest.java:
```diff
@@ -0,0 +1,47 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+public class StringAddCalculatorTest {
```
전체적인 테스트 케이스는 잘 나누어주신것 같아요. 
한가지 아쉬운 부분은 다양한 Parameter값에 대한 테스트가 없다는것, 테스트 네이밍을 Method 네이밍으로 지정해주신부분이 조금 아쉬운것 같아요.
@ParameterizedTest 와 @DisplayName을 활용해보는건 어떨까요 ?

sah3122 commented on src/test/java/study/calculator/StringAddCalculatorTest.java:
```diff
@@ -0,0 +1,47 @@
+        assertThatThrownBy(() -> StringAddCalculator.splitAndSum("-1,2,3"))
+                .isInstanceOf(RuntimeException.class);
+    }
+}
```
파일의 마지막은 NewLine으로 끝나는 것을 권장합니다. 
https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline

ljhokgo commented:
꼼꼼한 리뷰 너무 감사드립니다!
주신 코멘트 기반으로 수정해 보았습니다.

한가지 여쭤보고 싶은 것이 있습니다.
주신 코멘트 중

> null을 반환하는것은 좋지 않은 패턴입니다.
> Custom구분자가 아닐경우 Default구분자로 분리하는 로직을 null대신 사용해보세요!

요 부분을 반영해 split() 메소드를 개선해 보았는데요.

    private static String[] split(String input) {
        Matcher m = pattern.matcher(input);

        if (m.find()) {
            String customDelimiter = m.group(1);
            return m.group(2).split(customDelimiter);
        }
  
        return input.split(BASE_DELIMITER);
    }

커스텀 구분자가 있지 않더라도 무조건 Matcher 객체를 생성하게 되버려서..
더 나은 방법을 고민해봤는데 떠오르지가 않습니다..ㅠㅠ
이렇게 하는 것이 맞을까요..?

sah3122 commented on src/main/java/study/calculator/model/Number.java:
```diff
@@ -0,0 +1,27 @@
+        return value;
+    }
+
+    private int getIntFromString(String str) {
```
getXXX 보다는 convertXXX라는 네이밍이 어떨까요 ?
개인적으로 getXXX은 객체가 가지고 있는 상태를 가져올때 사용합니다 !

sah3122 commented on src/main/java/study/calculator/model/Number.java:
```diff
@@ -0,0 +1,27 @@
+
+        int number = Integer.parseInt(str);
+
+        if(number < 0) {
```
매직넘버는 상수로 선언해주세요 ! 
https://www.slipp.net/questions/356

sah3122 commented on src/main/java/study/calculator/model/Number.java:
```diff
@@ -0,0 +1,27 @@
+package study.calculator.model;
+
+public class Number {
+    private int value;
```
이왕이면 Number객체를 불변객체로 만들어 보는건 어떨까요 ?

sah3122 commented:
>커스텀 구분자가 있지 않더라도 무조건 Matcher 객체를 생성하게 되버려서..
더 나은 방법을 고민해봤는데 떠오르지가 않습니다..ㅠㅠ
이렇게 하는 것이 맞을까요..? 

처음과 같은 방식으로 커스텀 구분자의 유무를 판단한다면 커스텀 구분자의 패턴이 변경될 경우 수정해야하는 포인트가 늘어날것 같아요, 또한 완전한 Pattern을 검사해야 하기때문에 Matcher로 판단하는것이 더 나을것 같습니다 !
